Add physical server REST api

rodneyhbrown7 opened the pull request:
Add support for REST API operations against the physical server.   Includes
power on, power off, restart, loc LED on/off/blinking.

Depends on PR: https://github.com/ManageIQ/manageiq/pull/14026

@miq-bot add_label: wip, providers/physical_infrastructure

rodneyhbrown7 commented:
Closing and reopening pull request to re-run travis tests

rodneyhbrown7 commented:
@miq-bot add_label: providers/physical-infrastructure

rodneyhbrown7 commented:
Closing to rerun travis test with upstream resolutions

rodneyhbrown7 commented:
Re-open to run travis tests with upstream resolution

juliancheal commented on app/models/host.rb:
```diff
@@ -90,7 +90,10 @@ class Host < ApplicationRecord
-
+  
+  # Physical infra reference
+  has_one :physical_server, :foreign_key => "serial_number", :primary_key => "service_tag", :class_name => "PhysicalServer"
```
This is a repeat of the commits in https://github.com/ManageIQ/manageiq/pull/14026

juliancheal commented on app/models/physical_server.rb:
```diff
@@ -1,15 +1,44 @@
   belongs_to :ext_management_system, :foreign_key => :ems_id, :class_name => "ManageIQ::Providers::PhysicalInfraManager"
 
   default_value_for :enabled, true
+  has_one :host, :foreign_key => "service_tag", :primary_key => "serial_number"
```
This is a repeat of the commits in #14026

rodneyhbrown7 commented:
@Fryguy There are some rubocop complaints on this PR about naming is_refreshable to  refreshable? .  Can  those comments be ignored or do you recommend that we follow the suggested naming convention?

rodneyhbrown7 commented:
Re-open to rerun travis tests

rodneyhbrown7 commented:
Re-open to run travis tests

rodneyhbrown7 commented:
Close and re-open to run travis tests

rodneyhbrown7 commented:
Re-open to re-run travis tests

imtayadeway commented:
/cc @imtayadeway

rodneyhbrown7 commented:
@miq-bot add_label enhancement, providers/physical_infrastructure

rodneyhbrown7 commented:
@miq-bot remove_label wip

rodneyhbrown7 commented:
@miq-bot add_label providers/physical-infrastructure, api

walteraa commented:
Note: The current rubocop warnings is referring to methods which are used by other MIQ pieces, I don't believe is possible to change it.

walteraa commented:
@blomquisg @rodneyhbrown7 @juliancheal  Do you guys know why the merge cause erros in `ManageIQ::Providers::Openstack::CloudManager::Refresher` tests? The merge commit was https://github.com/ManageIQ/manageiq/pull/14028/commits/ae1ee1bc9c599629176084f7a1f00b1108047f61

rodneyhbrown7 commented:
Close and re-open to rerun travis tests.

rodneyhbrown7 commented:
Reopen to re-run travis tests

rodneyhbrown7 commented:
@blomquisg @imtayadeway @juliancheal  this PR is ready for review

imtayadeway commented on app/controllers/api/physical_servers_controller.rb:
```diff
@@ -1,4 +1,50 @@
 module Api
   class PhysicalServersController < BaseController
+    def blink_loc_led_resource(type, id, _data)
+      change_resource_state(:blink_loc_led, type, id)
```
@rodneyhbrown7 this is looking good! The only thing I want to suggest is that maybe here you could wrap these calls/responses in an asynchronous `api_action` like we do for other power operations (see https://github.com/ManageIQ/manageiq/blob/master/app/controllers/api/vms_controller.rb#L11-L22 for an example of this), that is, if these operations are doing things over the network and are potentially costly in terms of response time.

rodneyhbrown7 commented:
@miq-bot add_label fine/yes

rodneyhbrown7 commented:
@miq-bot add_label wip

rodneyhbrown7 commented:
@miq-bot remove_label wip

abellotti commented:
please rebase for api/physical_servers_controller.rb Thanks.

rodneyhbrown7 commented:
Updated based on the merge of https://github.com/ManageIQ/manageiq/pull/14430

blomquisg commented:
@imtayadeway if you can give me a :+1: on the API changes, this is good to go.

imtayadeway commented:
@blomquisg this looks basically good, but we're going to need some request specs exercising this code - I don't always trust my own :eyes: . @rodneyhbrown7 LMK if you need any help WRT this, also happy to pair if we can find some time to do that too if that helps.

rodneyhbrown7 commented:
@skovic has started working on some spec tests to cover this functionality. I will point him your way. I will see if we can get an initial WIP posted either today or Monday. I had him start by looking at the host api spec tests and factory instances as a base.

abellotti commented on app/controllers/api/physical_servers_controller.rb:
```diff
@@ -0,0 +1,50 @@
+
+      api_action(type, id) do |klass|
+        server = resource_search(id, type, klass)
+        api_log_info(" Processing reque:st to #{state} #{server_ident(server)}")
```
typo ?  
maybe change to:

``` 
Processing change resource state request to ...
```

abellotti commented on app/controllers/api/physical_servers_controller.rb:
```diff
@@ -0,0 +1,50 @@
+      api_action(type, id) do |klass|
+        server = resource_search(id, type, klass)
+        api_log_info(" Processing reque:st to #{state} #{server_ident(server)}")
+        desc = state.to_s
```
the desc should be more descriptive as it's used by action_result below.

```
#{server_ident(server)} changing state to #{state}
```

abellotti commented on app/controllers/api/physical_servers_controller.rb:
```diff
@@ -0,0 +1,50 @@
+        api_log_info(" Processing reque:st to #{state} #{server_ident(server)}")
+        desc = state.to_s
+        task_id = queue_object_action(server, desc, :method_name => state, :role => :ems_operations)
+        action_result(true, desc, :task_id => task_id)
```
missing the catch here:

```
api_action ...
   begin
      above stuff
   rescue => err
      action_result(false, err.to_s)
   end
end
```

abellotti commented on config/api.yml:
```diff
@@ -1064,6 +1064,42 @@
+    :verbs: *gp
+    :klass: PhysicalServer
+    :subcollections:
+    :collection_actions:
```
missing :get: section in collection and resource section with appropriate identifiers for listing (show_list) and fetching (show) a physical server.

also, should implement the query post action for bulk queries.

rodneyhbrown7 commented:
@abellotti @imtayadeway   Updated this PR based on comments and also added controller tests based on discussions this afternoon.

abellotti commented:
Thanks @rodneyhbrown7 for the update. 

Could you add tests for fetching multiple resources as well as the bulk query post action, both can go in collections_spec.rb. cookie cutters for that already there in that file.  Thanks.

rodneyhbrown7 commented:
@abellotti I have updated the collection_spec tests for /api/physical_servers.  I also noted that the tests added were testing the collection and not the resource, which was causing weird things when the action name was :read.    I updated the tests accordingly.

abellotti commented on spec/requests/api/collections_spec.rb:
```diff
@@ -314,6 +314,11 @@ def test_collection_bulk_query(collection, collection_url, klass, id = nil)
+
+    it 'query PhysicalServers' do
+      FactoryGirl.create(:physical_server)
+      test_collection_query(:physical_servers, "/api/physical_servers", PhysicalServer)
```
use physical_servers_url for second paramater

abellotti commented on spec/requests/api/collections_spec.rb:
```diff
@@ -580,5 +585,10 @@ def test_collection_bulk_query(collection, collection_url, klass, id = nil)
+
+    it 'bulk query PhysicalServers' do
+      FactoryGirl.create(:physical_server)
+      test_collection_bulk_query(:physical_servers, "/api/physical_servers", PhysicalServer)
```
use physical_servers_url for second parameter

abellotti commented on spec/requests/api/physical_servers_spec.rb:
```diff
@@ -0,0 +1,160 @@
+
+        api_basic_authorize action_identifier(:physical_servers, :read, :resource_actions, :get)
+
+        run_get "/api/physical_servers/1"
```
run_get physical_servers_url(1)  

you can also use the id generated by above FactoryGirl (without :id => 1), i.e. ps = FactoryGirl.create(...), then use physical_servers_url(ps.id)

abellotti commented on spec/requests/api/physical_servers_spec.rb:
```diff
@@ -0,0 +1,160 @@
+        api_basic_authorize action_identifier(:physical_servers, :power_on, :resource_actions, :post)
+
+        request = gen_request(:power_on)
+        run_post("/api/physical_servers/1", request)
```
can be one-liner,   run_post(physical_servers_url(id), gen_request(:power_on)

abellotti replied:
same comment for the following tests

abellotti commented:
Thanks @rodneyhbrown7 for making these updates, LGTM for API side,  @imtayadeway any other changes needed ?

imtayadeway commented on app/controllers/api/physical_servers_controller.rb:
```diff
@@ -0,0 +1,53 @@
+        begin
+          server = resource_search(id, type, klass)
+          desc = "Requested server state #{state} for #{server_ident(server)}"
+          api_log_info("desc")
```
Is this possibly supposed to be bare word `desc`?

rodneyhbrown7 replied:
I should think so. Let me adjust that.

imtayadeway commented on app/controllers/api/physical_servers_controller.rb:
```diff
@@ -0,0 +1,53 @@
+    end
+
+    #
+    # Name: power_on_resource
```
I'm not sure these comments are adding anything - could we remove or otherwise make more verbose to justify why it's needed?

imtayadeway commented:
@abellotti @rodneyhbrown7 API portion looks good apart from a couple of nits above. Great work!

rodneyhbrown7 commented:
Thanks @imtayadeway and @abellotti for your help polishing this up. I made the last few updates. Please let me know if you see anything else.

sergio-ocon commented:
@abellotti Should it be a candidate for Fine/yes?

abellotti commented:
I don't think this can be for Fine release as the dependent PRs that were merged are only on master as far as I can tell.

sergio-ocon commented:
Thanks @abellotti. Having it on master would be great.

abellotti commented:
ping @blomquisg 👍 on the API side.
